azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -40,81 +41,243 @@
+            # populates the '_extensions' property.
+            self._extensions = None
+            vm_settings = self._fetch_vm_settings()
+
+            # These "extended" properties come from additional HTTP requests to the URIs included in the basic goal state
+            self._hosting_env = None
+            self._shared_conf = None
+            self._certs = None
+            self._remote_access = None
+            self._extensions_goal_state = None
+
+            self._fetch_extended_goal_state(xml_text, xml_doc, vm_settings)
+
+        except Exception as exception:
+            # We don't log the error here since fetching the goal state is done every few seconds
+            raise ProtocolError(msg="Error fetching goal state", inner=exception)
+
+    @property
+    def timestamp(self):
+        return self._timestamp
+
+    @property
+    def incarnation(self):
+        return self._incarnation
+
+    @property
+    def container_id(self):
+        return self._container_id
+
+    @property
+    def role_instance_id(self):
+        return self._role_instance_id
+
+    @property
+    def role_config_name(self):
+        return self._role_config_name
+
+    @property
+    def extensions_goal_state(self):
+        return self._extensions_goal_state
+
+    @property
+    def certs(self):
+        return self._certs
+
+    @property
+    def hosting_env(self):
+        return self._hosting_env
+
+    @property
+    def shared_conf(self):
+        return self._shared_conf
+
+    @property
+    def remote_access(self):
+        return self._remote_access
+
+    @staticmethod
+    def update_host_plugin_headers(wire_client):
+        """
+        Updates the container ID and role config name that are send in the headers of HTTP requests to the HostGAPlugin
+        """
+        # Fetching the goal state updates the HostGAPlugin so simply trigger the request
+        GoalState._fetch_goal_state(wire_client)
+
+    def update(self, force_update=False):
+        """
+        Updates the current GoalState instance fetching values from the WireServer/HostGAPlugin as needed
+        """
+        xml_text, xml_doc = GoalState._fetch_goal_state(self._wire_client)
+
+        vm_settings = self._fetch_vm_settings(force_update=force_update)
+
+        if force_update or self._incarnation != findtext(xml_doc, "Incarnation"):
+            # update the extended goal state, using vm_settings for the extensions (unless they are None, then use extensionsConfig)
+            self._initialize_basic_properties(xml_doc)
+            self._fetch_extended_goal_state(xml_text, xml_doc,  vm_settings)
+        else:
+            # else just ensure the extensions are using the latest vm_settings
+            if vm_settings is not None:
+                self._extensions_goal_state = vm_settings
+
+    def save_to_history(self, data, file_name):
```

Comment:
This is just a temporary method while I do more refactoring on the methods to fetch the manifests